cli/azd/pkg/infra/azure_resource_manager_test.go

```diff
@@ -245,3 +248,96 @@ func TestGetDeploymentResourceOperationsWithNestedDeployments(t *testing.T) {
 	require.Equal(t, 1, subCalls)
 	require.Equal(t, 2, groupCalls)
 }
+
+type listResourceGroupFn = func(context.Context, string, *azcli.ListResourceGroupOptions) ([]azcli.AzCliResource, error)
+
+func TestFindResourceGroupForEnvironment(t *testing.T) {
+
+	t.Parallel()
+
+	// sequence combines multiple steps into one function that can be passed to ListResourceGroupFn. The first invocation
+	// of the returned function called the first step, the second invocation calls the second, and so on. If the returned
+	// function is called more times then there are steps, it panics.
+	sequence := func(steps ...listResourceGroupFn) listResourceGroupFn {
+		i := 0
+		return func(ctx context.Context, subscriptionId string, listOptions *azcli.ListResourceGroupOptions) ([]azcli.AzCliResource, error) {
+			if i >= len(steps) {
+				panic("no step registered")
+			}
+
+			v, err := steps[i](ctx, subscriptionId, listOptions)
+			i++
+			return v, err
+		}
+	}
+
+	tests := []struct {
+		name              string
+		rgsFromFilter     []string
+		rgsFromNoFilter   []string
+		expectedName      string
+		expectedErrorText string
+	}{
+		{"oneTaggedResourceGroup", []string{"custom-rg-name"}, nil, "custom-rg-name", ""},
```

Comment:
In #849 @vhvb1989 wrote:

> From the developer perspective, this is so much better than duplicating all test.
But, from a debugger perspective, this becomes an issue. If there's a test fail in the middle of the loop, you need to update the test array temporarily to run the affected test alone.
>
> During CI, you need to re-run all the loop, there's no way to run just one of the tests.
>
> So, I usually prefer the code duplication, just to help making tests unique.

Happy to go back - but surprised we wouldn't be able to use the `-run` flag of `go test` to solve these problems.  It allows running a single subtest.  In practice I expect we always end up running all the tests in the package so either approach would work.  I find this much nicer when scanning the entire file.

For the VS Code debugging experience I get that it's a little worse, but you can always comment out the block here in the initializer with just the case you care about.

Reply:
As part of responding to Wallace's feedback over what component to mock, I ended up having to do a bit of iteration on the tests.  I found myself just using VS Code to comment out scenarios and while a little frustrating, I think on balance I like it more than a bunch of individual test methods that call into a helper.  I feel like these table-driven tests are idiomatic and we are starting to get some more of them in the codebase, so I'm going to keep what I have here for now.  But I'm happy to go change this across the codebase if we think we get real value from having separate entry points.